### PR TITLE
Handle real segfaults in fuzz targets

### PR DESCRIPTION
**Pull request by br-lewis**
This adds handling for real segfaults in native code in our fuzz targets. Unfortunately, the only way to do that effectively was to have the C++ code in the fuzzer detect the error, print a message, write the crash file, and quit rather than returning control to JS so the result isn't as nice as a normal error but should be more informative than before.

### CODE REVIEW

File: tests/signal_handlers/SIGSEGV/fuzz.js [hunk 1]
```diff
@@ -36,3 +38,22 @@ module.exports.SIGSEGV_ASYNC = (data) => {
 	}
 	i++;
 };
+
+module.exports.NATIVE_SIGSEGV_SYNC = (data) => {
+	if (i === 1000) {
```
Comment (bertschneider) [inline tests/signal_handlers/SIGSEGV/fuzz.js:43]: As the "execute on Xs invocation" is now used in multiple tests a dedicated function to abstract that away could be useful.

File: packages/fuzzer/fuzzing_async.cpp [hunk 2]
```diff
@@ -66,6 +68,14 @@ using FinalizerDataType = void;
 
 TSFN gTSFN;
 
+const std::string SEGFAULT_ERROR_MESSAGE = "Segmentation fault found in fuzz target";
+
+std::jmp_buf errorBuffer;
+
+void ErrorSignalHandler(int signum) {
+  std::longjmp(errorBuffer, signum);
```
Comment (bertschneider) [inline packages/fuzzer/fuzzing_async.cpp:76]: Could you please add a comment describing when this handler will be called, where `longjmp` jumps to and why that's required?

File: packages/fuzzer/fuzzing_async.cpp [hunk 3]
```diff
@@ -107,6 +117,10 @@ void CallJsFuzzCallback(Napi::Env env, Napi::Function jsFuzzCallback,
   // thread and continue with the next invocation.
 
   try {
+    if (setjmp(errorBuffer) != 0) {
```
Comment (bertschneider) [inline packages/fuzzer/fuzzing_async.cpp:120]: Could you please add a comment describing why this is handled in the addon's main thread?

File: packages/fuzzer/fuzzing_async.cpp [hunk 4]
```diff
@@ -107,6 +117,10 @@ void CallJsFuzzCallback(Napi::Env env, Napi::Function jsFuzzCallback,
...
+      std::cerr << SEGFAULT_ERROR_MESSAGE << std::endl;
+      exit(139);
```
Comment (bertschneider) [inline packages/fuzzer/fuzzing_async.cpp:122]: Will this exit cause libFuzzer report a finding? Extract `139` in a more readable constant?

File: packages/fuzzer/fuzzing_sync.cpp [hunk 5]
```diff
@@ -36,9 +40,22 @@ std::optional<FuzzTargetInfo> gFuzzTarget;
...
+std::jmp_buf errorBuffer;
 } // namespace
 
-void sigintHandler(int signum) { gSignalStatus = signum; }
+void sigintHandler(int signum) {
+  gSignalStatus = signum;
+}
+
+// This handles signals that indicate an unrecoverable error (currently only segfaults).
+// Our handling of segfaults is odd because it avoids using our Javascript method to print and instead prints a message
```
Comment (bertschneider) [inline packages/fuzzer/fuzzing_sync.cpp:51]: These lines seem pretty long. Could you please run `npm run format:fix` in the root directory?

**Review by bertschneider — COMMENTED** [review]
Great that you got this working! 

A few, but only minor, comments, overall this PR looks good.

File: tests/signal_handlers/native-signal/package.json [hunk 6]
```diff
@@ -0,0 +1,24 @@
+{
+	"name": "native-signal",
+	"version": "1.0.0",
+	"description": "",
+	"main": "dist/index.js",
+	"scripts": {
+		"install": "npm run build",
```
[the remark this answers is not in the document]
Comment (br-lewis) [reply tests/signal_handlers/native-signal/package.json:7]: It's already ignored though so I'm not sure what you mean.

File: package.json [hunk 7]
```diff
@@ -60,7 +60,7 @@
 		"packages/*"
 	],
 	"lint-staged": {
-		"**/*": "prettier --write --ignore-unknown --allow-empty --loglevel debug"
+		"**/!(compile_commands.json)*": "prettier --write --ignore-unknown --allow-empty --log-level debug"
```
Comment (bertschneider) [inline package.json:63]: Could this be added in `.prettierignore`?

File: tests/signal_handlers/native-signal/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (bertschneider) [reply tests/signal_handlers/native-signal/package.json]: Hmm, you're right, nevermind. Somehow it showed up as unversioned in the IDE.

File: tests/signal_handlers/SIGSEGV/fuzz.js (hunk 1 above)
Comment (bertschneider) [reply tests/signal_handlers/SIGSEGV/fuzz.js]: Fair enough :+1:

**Review by bertschneider — APPROVED** [review]
Looks good, thx!

Please squash the commits before/on merge.

File: package.json (hunk 7 above)
Comment (br-lewis) [reply package.json:63]: You would think so but nope! Prettier doesn't have a problem when run on its own, the problem comes from `lint-staged`. `lint-staged` will use that pattern to find files but resolves the files to actual paths that it gives to prettier and prettier will, if explicitly given a path to a symlink, error out.

Comment (bertschneider) [reply package.json:63]: yikes

**Pull request merged**